### PR TITLE
feat(subscriptions): add setting to reattempt payment after final retry

### PR DESCRIPTION
**Pull request by chickenn00dle**
### All Submissions:

* [ ] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes https://app.asana.com/0/1208702687864922/1208702687865959/f

This PR adds a new 'On-hold duration' setting for subscriptions. This setting adds a new retry rule set to the specified number of days in WooCommerce > Settings > Subscriptions > On-hold Duration:

Screenshot 2024-11-25 at 15 53 34

### How to test the changes in this Pull Request:

1. As admin, go to WooCommerce > Settings > Subscriptions and scroll down. Confirm there is a new section added for Newspack subscriptions settings. Also confirm the On-hold Duration setting is present.
2. Select a non-zero value and save. Confirm the setting persists on reload
3. As a reader, purchase any subscription product
4. As admin, go to the edit subscriptions page for this subscription (WooCommerce > Subscriptions)
5. Follow these steps to manually trigger 5 payment retries: https://woocommerce.com/document/subscriptions/develop/failed-payment-retry/#section-7
6. Run one more time, verify a sixth job is scheduled set to run the number of days set in the On-hold Duration setting:
Screenshot 2024-11-20 at 17 06 12
7. Visit the edit subscription page for the subscription. Confirm the subscription is still on-hold. Also confirm the renewal order is still in a pending status:
Screenshot 2024-11-20 at 17 10 12
8. Run the scheduled action one more time. Confirm no new scheduled action is added (this should be the final retry rule)
9. Go back to the edit subscription page and verify the subscription is now expired and the renewal order status is set to failed:
Screenshot 2024-11-20 at 17 12 03
10. Set the On-hold duration setting to 0 and repeat the above steps. This time the fifth scheduled action should trigger the subscription to expire and the renewal order to fail.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by chickenn00dle** [comment]
Removing from review after feedback in 1208702687864922-as-1208702687865959/f

**Comment by chickenn00dle** [comment]
Ready for review again!

File: includes/plugins/class-woocommerce-subscriptions.php
Comment (leogermani) [inline includes/plugins/class-woocommerce-subscriptions.php]: have you considered simply adding one rule with `$on_hold_duration` as interval and `expired` as status, instead of two rules?

And if `$on_hold_duration` is zero, change the status of the last rule?

I ask this because I think it would look a little cleaner in the admin. In my case, my on hold period was 3 days:

image

It's probably because I'm rushing the scheduled actions, but the last action shows out of the order, as if it happened before. Also, there is one unnecessary additional attempt to process the payment right after one failed.

File: includes/plugins/class-woocommerce-subscriptions.php
Comment (leogermani) [inline includes/plugins/class-woocommerce-subscriptions.php]: What are we checking with `count( $retry_rules )` ? In which case there will be no rules? If another plugin is tweaking it? Or if retries are disabled? If the latter, is there a more direct check for that option?

**Review by leogermani — COMMENTED** [review]
Works like a charm! I also confirmed that this works with pre-existing subscriptions

I left two non blocking questions. But before we merge it, let's check if we are ok to shipping this or if we will want to give publishers a heads up and maybe hold until they are ready to receive this change.

File: includes/plugins/class-woocommerce-subscriptions.php
Comment (chickenn00dle) [reply includes/plugins/class-woocommerce-subscriptions.php]: > have you considered simply adding one rule with $on_hold_duration as interval and expired as status, instead of two rules?

The problem with this is these retry rules are parsed at the time of failed payment before the actual retry happens which means the status for a given retry rule will be applied at the time the rule is created and before the retry is actually triggered. So adding a single rule with `$on_hold_duration` as interval and `expired` as status would result in the subscription being set to `expired` on the fifth retry with the sixth retry attempt doomed to fail since the subscription has already been set to `expired`.

I hope this makes sense. Let me know if you'd like to huddle to talk it out.

File: includes/plugins/class-woocommerce-subscriptions.php
Comment (chickenn00dle) [reply includes/plugins/class-woocommerce-subscriptions.php]: > What are we checking with count( $retry_rules ) ? In which case there will be no rules? If another plugin is tweaking it? Or if retries are disabled? If the latter, is there a more direct check for that option?

I added the check mostly to account for the possibility of another plugin tweaking the rules. This particular hook would never be triggered if retries are disabled.

That said, I realize now that I had been assuming we enforce retries to be enabled for our sites. If this isn't the case, we would also need to force subscriptions to expire when this isn't enabled I think?

**Comment by chickenn00dle** [comment]
Thanks for the review @leogermani!

> let's check if we are ok to shipping this or if we will want to give publishers a heads up and maybe hold until they are ready to receive this change.

Should I maybe create a new epic branch for this?

File: includes/plugins/class-woocommerce-subscriptions.php
Comment (leogermani) [reply includes/plugins/class-woocommerce-subscriptions.php]: Ah, gotcha, that makes sense

**Comment by leogermani** [comment]
> Should I maybe create a new epic branch for this?

Let's ask Katie before.

Can I suggest one last thing? I see that in https://github.com/Automattic/newspack-plugin/pull/3568 you are creating the same file. WDYT of creating a folder instead and having one file for each feature.

**Comment by chickenn00dle** [comment]
> I see that in https://github.com/Automattic/newspack-plugin/pull/3568 you are creating the same file. WDYT of creating a folder instead and having one file for each feature.

IMO, this isn't necessary quite yet since we aren't adding too much and I feel a single subscriptions integration class is enough to house all of the logic for both `on-hold duration` and cancellation reason metadata, at least until we end up adding more functionality. 

That said, if you prefer we divide this further, I can make the change!

**Comment by leogermani** [comment]
> That said, if you prefer we divide this further, I can make the change!

I always prefer small files over large files, small methods over large methods, etc :) 

Also, it makes the history looks better when, in the future, we finally decide to break things into more than one file and all the git history is less accessible.

Why save on files? :)

**Comment by leogermani** [comment]
ah, and let's do a feature flag instead of an epic? based on our recent experiences? wdyt?

**Comment by chickenn00dle** [comment]
> Also, it makes the history looks better when, in the future, we finally decide to break things into more than one file and all the git history is less accessible.

Sounds good!

> ah, and let's do a feature flag instead of an epic? based on our recent experiences? wdyt?

Great idea!

I've reorganized the code and added a feature flag in https://github.com/Automattic/newspack-plugin/pull/3560/commits/612c960fdc2590267d2ddfb52566d1b3daae8b41

Note I've also added a check for reader activation being enabled which I realized I forgot to include in this PR. 

Let me know if this works!

File: includes/plugins/class-woocommerce-subscriptions.php
Comment (leogermani) [reply includes/plugins/class-woocommerce-subscriptions.php]: > If this isn't the case, we would also need to force subscriptions to expire when this isn't enabled I think?

I don't think we enforce that. That's a good question. Let's open a ticket for that, as Katie and handle this in another PR if needed

**Review by leogermani — APPROVED** [review]

File: includes/plugins/woocommerce-subscriptions/class-woocommerce-subscriptions.php
Comment (leogermani) [inline includes/plugins/woocommerce-subscriptions/class-woocommerce-subscriptions.php]: you also need to check if the value of the constant is true? It can be defined as false

**Review by leogermani — CHANGES_REQUESTED** [review]

File: includes/plugins/woocommerce-subscriptions/class-woocommerce-subscriptions.php
Comment (chickenn00dle) [reply includes/plugins/woocommerce-subscriptions/class-woocommerce-subscriptions.php]: Good catch! Added in https://github.com/Automattic/newspack-plugin/pull/3560/commits/31f273ce720da5d2572bb100d39055e7dc6d8666

File: includes/plugins/woocommerce-subscriptions/class-woocommerce-subscriptions.php
Comment (leogermani) [inline includes/plugins/woocommerce-subscriptions/class-woocommerce-subscriptions.php]: this check doesn't look right.

`defined( self::NEWSPACK_SUBSCRIPTIONS_EXPIRATION_FEATURE_FLAG )` looks ok, because the class constant is a string and it will check for a global constant named after that string

but `self::NEWSPACK_SUBSCRIPTIONS_EXPIRATION_FEATURE_FLAG` will resolve as a string and thus always return `true`. 

Any reason to have this proxy constant in the class and not check the global constant directly?

If you want to keep it (no strong opinions here), I think you need to use `constant( self::NEWSPACK_SUBSCRIPTIONS_EXPIRATION_FEATURE_FLAG )`

Comment (chickenn00dle) [reply includes/plugins/woocommerce-subscriptions/class-woocommerce-subscriptions.php]: > but self::NEWSPACK_SUBSCRIPTIONS_EXPIRATION_FEATURE_FLAG will resolve as a string and thus always return true.

Ah. Whoops! That's my bad. Shouldn't have rushed that last commit.

> Any reason to have this proxy constant in the class and not check the global constant directly?

Not really, aside from explicitly defining the feature flag in the code for search-ability and future reference. That said, its fairly self-contained so removed in https://github.com/Automattic/newspack-plugin/pull/3560/commits/89dea71ea838b1dce1ecb05fd9e2d899fa97ef9c

**Review by leogermani — APPROVED** [review]

File: includes/plugins/woocommerce-subscriptions/class-on-hold-duration.php
Comment (claudiulodro) [inline includes/plugins/woocommerce-subscriptions/class-on-hold-duration.php:99]: Just double checking, if I've read this right it looks like the expiration date will be the date of the last failed payment retry? If so, that is the behavior publishers want!

Comment (chickenn00dle) [reply includes/plugins/woocommerce-subscriptions/class-on-hold-duration.php:99]: That's correct! Sorry for the late reply. Totally missed this comment.

**Pull request merged**
